massa-factory-worker/src/denunciation_factory.rs

```diff
@@ -0,0 +1,260 @@
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::thread;
+
+use crossbeam_channel::{select, Receiver};
+use tracing::{debug, info, warn};
+
+use massa_factory_exports::{FactoryChannels, FactoryConfig};
+use massa_models::block_header::SecuredHeader;
+use massa_models::denunciation::Denunciation;
+use massa_models::endorsement::SecureShareEndorsement;
+use massa_models::slot::Slot;
+use massa_models::timeslots::get_closest_slot_to_timestamp;
+use massa_time::MassaTime;
+
+// TODO: rework these values
+const DENUNCIATION_FACTORY_ENDORSEMENT_CACHE_MAX_LEN: usize = 4096;
+const DENUNCIATION_FACTORY_BLOCK_HEADER_CACHE_MAX_LEN: usize = 4096;
+
+/// Structure gathering all elements needed by the factory thread
+#[allow(dead_code)]
+pub(crate) struct DenunciationFactoryWorker {
+    cfg: FactoryConfig,
+    channels: FactoryChannels,
+    factory_receiver: Receiver<()>,
+    consensus_receiver: Receiver<SecuredHeader>,
+    endorsement_pool_receiver: Receiver<SecureShareEndorsement>,
+
+    // TODO: optim with PreHashSet?
+    // internal for endorsements (or secure share endorsements)
+    /// Internal storage for endorsement -
+    /// store at max 1 endorsement per entry, as soon as we have 2 we produce a Denunciation
+    endorsements_by_slot_index: HashMap<(Slot, u32), Vec<SecureShareEndorsement>>,
+
+    // internal for block header (or secured header)
+    /// Internal storage for endorsement -
+    /// store at max 1 endorsement per entry, as soon as we have 2 we produce a Denunciation
+    // FIXME: Store per 'Slot' then per 'PubKey'
+    block_header_by_slot: HashMap<Slot, Vec<SecuredHeader>>,
+}
+
+impl DenunciationFactoryWorker {
+    /// Creates the `FactoryThread` structure to gather all data and references
+    /// needed by the factory worker thread.
+    pub(crate) fn spawn(
+        cfg: FactoryConfig,
+        channels: FactoryChannels,
+        factory_receiver: Receiver<()>,
+        consensus_receiver: Receiver<SecuredHeader>,
+        endorsement_pool_receiver: Receiver<SecureShareEndorsement>,
+    ) -> thread::JoinHandle<()> {
+        thread::Builder::new()
+            .name("denunciation-factory".into())
+            .spawn(|| {
+                let mut factory = Self {
+                    cfg,
+                    channels,
+                    factory_receiver,
+                    consensus_receiver,
+                    endorsement_pool_receiver,
+                    endorsements_by_slot_index: Default::default(),
+                    block_header_by_slot: Default::default(),
+                };
+                factory.run();
+            })
+            .expect("failed to spawn thread : denunciation-factory")
+    }
+
+    /// Process new secured header (~ block header)
+    fn process_new_secured_header(&mut self, secured_header: SecuredHeader) {
+        let key = secured_header.content.slot;
+
```

Comment:
If the public key of the header creator is not checked, someone can spam with headers coming from non-PoS-drawn pubkeys and cause a problem

Reply:
done